### ISSUE TITLE
Adnan Ul-Hasan的博士论文-第七章  多种文字确认的序列学习

### ISSUE THREAD

**Issue by wanghaisheng** [wanghaisheng/awesome-ocr#83]
Script identification is a crucial step in digitizing multilingual documents. In addition
to simplifying the OCR process, it could be used for many other tasks like document
indexing, routing, and categorization of documents [PD14]. The traditional process
to OCR documents with multiple scripts is to identify and separate the scripts before
applying the OCR models. By doing so, the problem is reduced to applying singlescript
OCR models to the separated scripts.
A standard script identification approach is to manually extract salient features
of the given script(s) and then apply a pattern classification algorithm. However, contemporary
ML techniques can learn the distinguishing features automatically; thereby
eliminating the manual feature extraction step. Sequence learning is based on the
sequence-to-sequence matching principle where a neural network is trained such that
the input-output pair is comprised of sequences (text-lines in the case of documents).
The contribution of this thesis1 in using sequence learning for multiple script identification
is described in this chapter. The highlights of the proposed approach are:
• No script-specific handcrafted features are extracted. The LSTM networks have
yielded excellent results using only the raw pixel values.
• No word- or character-segmentation is required (text-lines extraction is however
essential); yet the script identification is done at character level.
• The LSTM networks are used for sequence-to-sequence mapping.

**Comment by wanghaisheng** [comment]
This chapter is further organized as follows: Section 7.1 describes the scope of
this chapter with respect to the categorization of multilingual documents given in
Chapter 3. Sections 7.2 and 7.3 report various script identification techniques used
by other researchers. Section 7.4 details the LSTM-based sequence-learning technique
for multiple script identification. Section 7.5 describes the experimental setup,
database used, and performance metrics; while Section 7.6 shows the results of experimental
evaluation and discusses the errors in detail. Section 7.7 concludes the
chapter with a short summary suggesting few possible future directions in which this
work can be extended.
7.1 Scope
A general categorization of multilingual documents has been described in Chapter 3.
The first class of multilingual documents are those which have different languages
sharing the same script, while the second type are those in which a single language
is written in two or more scripts. The documents in the third category contain different
languages and different scripts, for instance, English-Arabic, English-Greek, and
Chinese-Urdu. For such documents, visual appearance of underlying scripts is highly
discriminative and multiple scripts can be separated more easily as compared to the
documents in other two categories.
This chapter demonstrates the capacity of LSTM networks in simplifying the script
identification task by 1) removing the feature extraction step, and 2) assigning script
labels at character level despite processing the whole text-line.
7.2 Heuristic-Based Approaches for Script Identification
Pioneering work on multiple script identification was done by Spitz [Spi97]. They used
optical density distribution as a feature for the Han script and used typographic character
codes for Latin-based script identification. In an another early work, Pal and
Chaudhuri [PC02] developed a script identification system to recognize printed Roman,
Chinese, Arabic, Devanagari, and Bangla text lines. They used different features
to identify a particular script.
Ghosh et al., in [GDS10], reported a comprehensive review on multiple script recognition.
They divided the script recognition process into two categories of structureand
appearance-based approaches and provided comparative analysis of numerous
systems. Appearance-based approaches are global in the sense that they extract textblocks
irrespective of the document’s script. These techniques are usually faster and
are applicable to a wide variety of scripts. Structure-based are local approaches, which
are useful where script identification is required at line-, word- or character-level.
Both approaches require segmentation of text blocks at the preprocessing stage.
Wang et al. [WLS10] proposed a noise tolerant script identification system based
on pixel distribution of low-resolution skeleton of a character or a sub-word. This feature
discriminated Chinese, Korean, Japanese, and English scripts at connected component
level. Padma and Vijaya [PV10] proposed a texture-based global approach
based on Wavelet packet feature (Haar basis function). They reported the script identification
accuracies for seven scripts used in India.
Sharma et al. [SCPB13] proposed a methodology for script identification in video
frames. They used three well-known features, namely, Zernike moments, Gabor, and
gradient features. A script was recognized using Support Vector Machines (SVM) classifier
on these features. They evaluated their method on English, Hindi, and Bangla
scripts. In order to overcome the problems associated with video frames, they used
super-resolution and skeletonization as preprocessing steps.
Rani et al. [RDL13] proposed a script identification method for pre-segmented English
and Gurmukhi characters. They extracted Gabor features and gradient features
from the characters and applied SVM classifier for recognition.
Recently, Echi et al. [ESB14] proposed a script identification technique for both
printed and handwritten text for Arabic-Latin documents. They used many known
features in addition to their own features to identify Arabic and Latin. They compared
the performance of several classifiers including Naive Bayes, k-NN, Decision Trees,
Multilayer perceptron (MLP) and SVM on selected features.
All of the above-mentioned methods rely on extraction of sophisticated features
and image processing techniques along with various heuristics. This is because of the
common belief that script-related features are essential to be extracted for reliable
script identification. Therefore, major effort has been spent by the research community
to find and fine-tune script-related features. One major issue with all featurebased
methodologies is the extensive use of heuristics to adapt these methods to
the task at hand. Therefore, techniques designed for one script or a set of scripts are
not generally suitable for other scripts.
7.3 Machine Learning Based Approaches
Contemporary Machine Learning (ML) techniques, like Deep Belief Networks (DBN)
and deep Convolutional Neural Networks (CNN), are gaining popularity in the fields
of Computer Vision (CV) and Pattern Recognition (PR), and one of the reason for their
wide application in this domain is that they do not rely on sophisticated handcrafted
features. Instead they learn the features from the given data themselves, and still
perform better than the methods based on handcrafted features. Likewise, there has
been some recent efforts in applying these techniques for script identification.
Rashid et al. [Ras10] proposed a script identification method based on CNN, which
learns the shapes of individual connected components without explicitly extracting
features from them. They tested their method on Latin/Greek, Latin/Arabic and Antiqua/
Fraktur scripts. However, they applied a post-processing step to assign small
diacritics to corresponding base shapes to improve the final accuracy.
Genzal et al. [GPTF13] proposed an HMM-based script identification system. On a
private database consisting of 54 languages with 18 different scripts, they reported
an average script identification error rate of 1.73%. The text-lines in their work were
assigned a dominant script if it had more than one script.
Singh and Jawahar [SJ15] recently proposed an RNN-based approach for script
and language separation. Their approach identifies words of different scripts and languages
in a text-line. They cut the word image into upper and lower half for profile
feature extraction. The result on a large corpus containing 15 languages shows an
average accuracy of 93.96% at word-level and 97.9% at text-line level.
Although all of the above-mentioned techniques are based on modern machine
learning principles, but the first work applies post-processing step to improve the accuracy,
the second approach achieves low recognition error by the use of carefully
selected features and the third work employs profile features of words for the recognition.
Our approach, however, is very simple as we neither extract any feature(s) nor
we have applied any kind of post-processing on the output. The details of this method
are described in the following section.
7.4 LSTM Networks for Script Identification
The scheme introduced in this chapter models the script identification as a multi-class
classification problem. A single target class-label is assigned to all characters of a particular
script, which the LSTM network learns in a supervised classification paradigm.
A text-line is not segmented into individual characters; instead LSTM network learns
scripts’ shapes in a sequence-to-sequence mapping manner. In other words, we are
training the LSTM network to learn a particular target class-label for all alphabets in
a given language/script.
Figure 7.1 shows the proposed script identification process. The open source
OCRopus document analysis system [OCR15] is used for the experimental evaluation.
The OCRopus line recognizer is basically designed for OCR task, where the goal is
to transcribe a given text-line; therefore, the proposed idea essentially adapts this
OCR system to perform the script recognition task. The modification made in the
current design to adapt the OCR system is very simple. Instead of giving the Ground-
Truth (GT) information about the individual characters in a text-line, a sequence of
class-labels is provided as the GT at the training stage . To achieve this in practice, the
GT information is modified such that it represents a two-class classification problem
(see Figure 7.1). The modified GT information is used as the target sequence for a
given text-line image during the training. LSTM network learns the class association
of each characters based on their own shape as well as the contextual information of
7.5 Experiments
7.5.1 LSTM Networks’ Architecture
For the experiments reported in this chapter, 1D-LSTM architecture has been employed.
It is the same as described in [BUHAAS13]. In 1D-LSTM networks, a sliding
window of 1-pixel width and of height equal to the image height, traverses the input
text-line image to convert it into a one dimensional sequence. The height of the
image is termed as “depth” of the 1D sequence and each slice of the image thus obtained
is called a “frame”. It is necessary to make the depth of all images equal, so
that the individual frames in each image are equal. This process of making heights
equal is labelled as “normalization”, and it is essential for such types of networks. For
this work, we use the same normalization method used in [YSBS15] and it is available
in the OCRopus system. This method of normalization, which can be used for many
other scripts, is based on filter operations and affine transformations.
The LSTM network consists of one input layer, one hidden layer containing 50 LSTM
cells and one output layer. For 1D-LSTM training, a sliding window of 48 × 1 scans the
input image and each frame is given to the LSTM network for training. The network
learns to classify each frame into one of the target classes, including the reject class,
based on the contextual information of that frame. The bidirectional mode of LSTM
architecture is used to access the context in both forward and backward direction. In
bidirectional mode, two hidden layers scan the input. One traverses the input from
right-to-left and the other from left-to-right and both layers are connected to a single
output layer.
7.5.2 Database
There is no standard database available for script identification purpose. Therefore, a
synthetic database consisting of text-line images from freely available English-Greek
text has been developed to evaluate the proposed method. A text-line generation
tool available in OCRopus has been used for this purpose. This tool can generate any
number of text-line images with a given text-file and font-files. A modification in this
utility is done to generate the GT data for each transcription that suits our needs.
90,000 text-line images are synthetically generated in a variety of fonts (both serif
and sans-serif) using degradation models described in Section 4.3. A separate test
database consisting of 9,500 text-line images from a different English-Greek text has
also been prepared. This database is available free of cost for the research purposes
from the author.
7.5.3 Performance Metric
Since we have adapted an OCR system for the script identification purpose, the output
is like a transcription. Therefore, the performance is measured in terms of edit
distance (or Levenshtein Distance). Section 5.1.3 describes the details on this metric
7.6 Results and Discussions
There are a total of 9,500 text-line images in our test dataset consisting of 289,013
characters. LSTM network trained for script-identification task as described in Section
7.4 yielded an accuracy of 98.19% (an error rate of 1.81%). Some sample test
images are shown in Figure 7.3. Top 15 confusions are shown in Table 7.1. It should
From Table 7.1, one can see that many top confusions are due to the similarity in
shapes of characters between the two scripts, like ‘v’ with 􀀸, ‘o’ with 􀀺, ‘i’ with 􀀴, ‘p’ with
􀀼, ‘a’ with 􀀬, ‘u’ with 􀁀, ‘t’ with 􀀿 and so on. These type of errors will not be present
when characters of both scripts are significantly different, e.g., English and Hindi or
Arabic and Chinese. A bigram or trigram analysis could be done to further investigate
the causes of such confusions.
Many confusions occur at the locations where the script of the document changes
from English to Greek or vice versa. For example, in Figure 7.2, the second last character
􀁂 is misclassified as ‘X’ because the preceding character is an ‘r’ (English). Contextual
information at that time-step incorrectly makes the LSTM network classify this
character as belonging to the Latin script. Again, such errors will not occur in documents
in which character shapes are considerably different.
Script identification has been reported by many researchers, but the unavailability
of a standard database hampers the comparison of true performance among reported
methodologies. Moreover, the difference in performance metrics also obstructs a fair
comparison.
The results also show the powerful learning capabilities of LSTM networks. LSTM
networks are able to learn many variation in shapes for the same class, for example,
it could recognize the class association of whole English and Greek scripts to a single
class-label. This suggests us that these networks are very useful for large-scale
classification problems where class members show large variations.
7.7 Chapter Summary
This chapter introduces a new method for identifying multiple scripts in multilingual
documents. LSTM-based line recognizer has been trained to learn class associations
of individual characters for a given bilingual text-line. Experimental evaluation on
bilingual English-Greek data has shown encouraging results. This work can be extended
further in numerous ways. Firstly, other bilingual documents such as English-
Devanagari, Arabic-Latin, or English-Chinese can directly benefit from this approach.
Secondly, this approach can be adapted to documents with more than two scripts.
Another direction in which the research can be extended is that the LSTM networks
can be used to directly OCR the multilingual documents. They have demonstrated a
good capacity to learn a lot of classes at the same time; this means that we can avoid
the script identification step in some, if not all, multilingual documents.
